### PR TITLE
Functional tests for task specific and shared Docker volumes

### PR DESCRIPTION
**Pull request by sharanyad**
### Summary
Functional tests for local and shared docker volumes for tasks

### Implementation details
Note: This PR cannot be merged until backend changes are in prod. 
TODO 
- [ ] Verify shared volumes not removed on task cleanup (only done manually for now)
- [ ] Shared, auto provisioned
- [ ] Windows tests

### Testing
- [x] Builds on Linux (`make release`)
- [x] Builds on Windows (`go build -out amazon-ecs-agent.exe ./agent`)
- [x] Unit tests on Linux (`make test`) pass
- [x] Unit tests on Windows (`go test -timeout=25s ./agent/...`) pass
- [ ] Integration tests on Linux (`make run-integ-tests`) pass
- [ ] Integration tests on Windows (`.\scripts\run-integ-tests.ps1`) pass
- [x] Functional tests on Linux (`make run-functional-tests`) pass
- [ ] Functional tests on Windows (`.\scripts\run-functional-tests.ps1`) pass

New tests cover the changes: yes

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: agent/ecs_client/model/api/api-2.json [hunk 1]
```diff
@@ -679,6 +685,7 @@
       "members":{
         "name":{"shape":"String"},
         "image":{"shape":"String"},
+        "repositoryCredentials":{"shape":"RepositoryCredentials"},
```
Comment (yunhee-l) [inline agent/ecs_client/model/api/api-2.json]: is this mixed with private registry work? if so, is this blocking and/or cause conflicts with any other work?

File: agent/ecs_client/model/api/api-2.json [hunk 2]
```diff
@@ -807,7 +814,8 @@
         "placementConstraints":{"shape":"PlacementConstraints"},
         "placementStrategy":{"shape":"PlacementStrategies"},
         "networkConfiguration":{"shape":"NetworkConfiguration"},
-        "healthCheckGracePeriodSeconds":{"shape":"BoxedInteger"}
+        "healthCheckGracePeriodSeconds":{"shape":"BoxedInteger"},
+        "schedulingStrategy":{"shape":"SchedulingStrategy"}
```
Comment (yunhee-l) [inline agent/ecs_client/model/api/api-2.json]: this doesn't seem related to volume plugin work

File: agent/ecs_client/model/api/api-2.json (hunk 1 above)
Comment (adnxn) [reply agent/ecs_client/model/api/api-2.json]: looks like this change is just bringing in the acs model changes. it shouldnt block and im expecting it to just resolve itself during merge.

File: agent/ecs_client/model/api/api-2.json (hunk 2 above)
Comment (adnxn) [inline agent/ecs_client/model/api/api-2.json]: are we expecting to bring these in already? specifically `schedulingStrategy`?

File: agent/ecs_client/model/api/api-2.json [hunk 3]
```diff
@@ -1475,6 +1507,13 @@
         "taskDefinition":{"shape":"TaskDefinition"}
       }
     },
+    "RepositoryCredentials":{
+      "type":"structure",
+      "required":["credentialsParameter"],
+      "members":{
+        "credentialsParameter":{"shape":"String"}
```
Comment (tehshwann) [inline agent/ecs_client/model/api/api-2.json]: Does this allow for more than one credential set? And is that something we should consider?

File: agent/functional_tests/tests/generated/simpletests_unix/simpletests_generated_unix_test.go [hunk 4]
```diff
@@ -762,6 +762,86 @@ func TestSimpleExit(t *testing.T) {
...
+		if exit, ok := testTask.ContainerExitcode("exit"); !ok || exit != 42 {
+			t.Errorf("Expected exit to exit with 42; actually exited (%v) with %v", ok, exit)
+		}
+
+		defer agent.SweepTask(testTask)
+	}
+
+}
+
+// TestTaskSharedVolume Verify that shared Docker volume works as expected
```
Comment (tehshwann) [inline agent/functional_tests/tests/generated/simpletests_unix/simpletests_generated_unix_test.go]: Should we add some tests here that write real data to the volume rather than just "touching" it? For example... perhaps we should have something that fills the volume to make sure it can actually be used by a task?

Comment (richardpen) [reply agent/functional_tests/tests/generated/simpletests_unix/simpletests_generated_unix_test.go]: A better way to test the shared volume may be actually starting two tasks:
* task one to write to the volume
* task two read from the volume and verify the content
Any thoughts?

Comment (sharanyad) [reply agent/functional_tests/tests/generated/simpletests_unix/simpletests_generated_unix_test.go]: That's a good idea. I think we could test for volume cleanup through this too, where task 1 writes to the volume and stops executing (volume not removed here), and task 2 reads from the same volume.

**Comment by sharanyad** [comment]
Closing this in favor of a different PR that @yunhee-l will put up.
